CSharp/Library/Microsoft.Bot.Connector.Shared/Authentication/AdalAuthenticator.cs

```diff
@@ -0,0 +1,115 @@
+            authContext.TokenCache.Clear();
+        }
+
+        private async Task<AuthenticationResult> AcquireTokenAsync()
+        {
+            bool acquired = false;
+
+            try
+            {
+                // The ADAL client team recommends limiting concurrency of calls. When the Token is in cache there is never 
+                // contention on this semaphore, but when tokens expire there is some. However, after measuring performance
+                // with and without the semaphore (and different configs for the semaphore), not limiting concurrency actually
+                // results in higher response times overall. Without the use of this semaphore calls to AcquireTokenAsync can take up
+                // to 5 seconds under high concurrency scenarios.
+                acquired = await authContextSemaphore.WaitAsync(SemaphoreTimeout).ConfigureAwait(false);
+
+                // If we are allowed to enter the semaphore, acquire the token.
+                if (acquired)
+                {
+                    // Acquire token async using ADAL.NET
+                    // https://github.com/AzureAD/azure-activedirectory-library-for-dotnet
+                    // Given that this is a ClientCredential scenario, it will use the cache without the 
+                    // need to call AcquireTokenSilentAsync (which is only for user credentials).
+                    var res = await authContext.AcquireTokenAsync(JwtConfig.OAuthResourceUri, clientCredential).ConfigureAwait(false);
+                    return res;
+                }
+
+                throw new TimeoutException("Timeout waiting for token acquisition.");
+            }
+            finally
+            {
+                // Always release the semaphore.
+                if (acquired)
+                {
+                    authContextSemaphore.Release();
+                }
+
+            }
+        }
+
+        private RetryParams HandleAdalException(Exception ex, int currentRetryCount)
+        {
+            if (ex is AdalServiceException)
+            {
+                AdalServiceException adalServiceException = ex as AdalServiceException;
+
+                // When the Service Token Server (STS) is too busy because of “too many requests”, 
+                // it returns an HTTP error 429 with a hint about when you can try again (Retry-After response field) as a delay in seconds
+                if (adalServiceException.ErrorCode == AdalError.ServiceUnavailable || adalServiceException.StatusCode == 429)
+                {
+                    RetryConditionHeaderValue retryAfter = adalServiceException.Headers.RetryAfter;
+
+                    // Depending on the service, the recommended retry time may be in retryAfter.Delta or retryAfter.Date. Check both.
+                    if (retryAfter != null && retryAfter.Delta.HasValue)
+                    {
+                        return new RetryParams(retryAfter.Delta.Value);
```

Comment:
OK I see how you bound this later.

There are two ways you can do this:
* The structure carries the retry *request* (i.e. what the service gave you)
* The structure carries the retry *decision* (i.e. what you chose the retry interval to be after applying bounds and noise)

I think the latter is clearer. That is, make all decisions about the retry interval within this function. When the function returns, the retry interval is 100% complete.

Reply:
Ha, good way of expressing it. I feel like now the structure carries the retry decision. Should I clarify further in comments? Or are there any changes you have in mind here?

Reply:
Btw I made a small change, if the server sends a 429 with too long of a retry after header (i.e. 10 seconds) instead of throwing (which doesn't help anybody) we'll retry at maxDelay (See line 32 after I push)